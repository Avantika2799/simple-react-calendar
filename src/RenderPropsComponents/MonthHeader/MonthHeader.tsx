import React, { FC, ReactElement } from 'react'

type Props = {
  blockClassName?: string
  onMonthChange: HandleOnMonthChange
  ComponentMonthTitle: ReactElement
  ComponentButtonPrev?: ReactElement
  ComponentButtonNext?: ReactElement
  ComponentMonthDays?: ReactElement
}

const displayName = 'MonthHeader'

const MonthHeader: FC<Props> = ({ blockClassName, ComponentMonthTitle, ComponentButtonPrev, ComponentButtonNext, ComponentMonthDays }) => {
  const className = `${blockClassName}-${displayName}`

  return (
  <div className={className}>
    <div className={`${className}-top`}>
    {ComponentButtonPrev}
    <div className={`${className}-title`}>
      {ComponentMonthTitle}
    </div>
    {ComponentButtonNext}
    </div>
    {ComponentMonthDays && <div className={`${className}-bottom`}>
      {ComponentMonthDays}
                           </div>}
  </div>)
}

MonthHeader.displayName = displayName

export default MonthHeader
