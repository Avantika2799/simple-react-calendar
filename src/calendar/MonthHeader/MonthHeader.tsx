import React, { FC, SyntheticEvent } from 'react'

import { IMonthHeaderButtonRenderProps, IMonthHeaderDayRenderProps, IMonthHeaderRenderProps } from '../../@types'
import MonthHeaderButton from '../MonthHeaderButton'
import MonthHeaderDay from '../MonthHeaderDay'

export type Props = {
  activeMonth: Date
  blockClassName: string
  customRender?: IMonthHeaderRenderProps
  getMonthFormattedName: (day: Date) => string
  getWeekDays: () => string[]
  getDayIndexIsWeekend: (index: 0 | 1 | 2 | 3 | 4 | 5 | 6) => boolean
  handleOnSwitchMonth: (event: SyntheticEvent<HTMLElement>) => void
  isMonthNextEnabled: boolean
  isMonthPrevEnabled: boolean
  renderMonthHeaderButton?: IMonthHeaderButtonRenderProps
  renderMonthHeaderDay?: IMonthHeaderDayRenderProps
  showDaysOfWeek: boolean
  showMonthName: boolean
  showMonthNext: boolean
  showMonthPrev: boolean
}

const MonthHeader: FC<Props> = (props) => {
  const {
    activeMonth,
    blockClassName,
    customRender,
    getMonthFormattedName,
    getWeekDays,
    getDayIndexIsWeekend,
    handleOnSwitchMonth,
    isMonthNextEnabled,
    isMonthPrevEnabled,
    renderMonthHeaderButton,
    renderMonthHeaderDay,
    showDaysOfWeek,
    showMonthName,
    showMonthNext,
    showMonthPrev,
  } = props

  if (customRender) {
    return customRender(props)
  }

  if (!showMonthPrev && !showMonthName && !showMonthNext && !showDaysOfWeek) return null

  return (
    <header className={`${blockClassName}-month_header`}>
      <div className={`${blockClassName}-month_header_wrapper`}>
        {showMonthPrev && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthPrevEnabled}
            handleOnClick={handleOnSwitchMonth}
            type='prev'
          />
        )}
        {showMonthName && (
          <h3 className={`${blockClassName}-month_header_title`}>{getMonthFormattedName(activeMonth)}</h3>
        )}
        {showMonthNext && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthNextEnabled}
            handleOnClick={handleOnSwitchMonth}
            type='next'
          />
        )}
      </div>
      <div className={`${blockClassName}-month_header_days`}>
        {showDaysOfWeek &&
          getWeekDays().map((day, index) => (
            <MonthHeaderDay
              blockClassName={blockClassName}
              customRender={renderMonthHeaderDay}
              day={day}
              isWeekend={getDayIndexIsWeekend(index)}
              key={day}
            />
          ))}
      </div>
    </header>
  )
}

MonthHeader.displayName = 'MonthHeader'

export default MonthHeader
