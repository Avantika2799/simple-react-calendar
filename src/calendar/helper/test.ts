import { getDayEachDay, getDayIndexIsWeekend, getMonthWeeks } from '.'

describe('Helper functions', () => {
  describe('#getDayIndexIsWeekend', () => {
    describe('when weekday', () => {
      describe('when `index` is `Monday`', () => {
        test('returns `false`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 1, weekStartsOn: 1 })).toBe(false)
        })
      })

      describe('when `index` is `Tuesday`', () => {
        test('returns `false`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 2, weekStartsOn: 1 })).toBe(false)
        })
      })

      describe('when `index` is `Wednesday`', () => {
        test('returns `false`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 3, weekStartsOn: 1 })).toBe(false)
        })
      })

      describe('when `index` is `Thursday`', () => {
        test('returns `false`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 4, weekStartsOn: 1 })).toBe(false)
        })
      })

      describe('when `index` is `Friday`', () => {
        test('returns `false`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 5, weekStartsOn: 1 })).toBe(false)
        })
      })
    })

    describe('when weekend', () => {
      describe('when `index` is `Saturday`', () => {
        test('returns `true`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 6, weekStartsOn: 1 })).toBe(true)
        })
      })

      describe('when `index` is `Sunday`', () => {
        test('returns `true`', () => {
          expect(getDayIndexIsWeekend({ dayIndex: 0, weekStartsOn: 1 })).toBe(true)
        })
      })
    })
  })

  describe('#getDayEachDay', () => {
    test('returns array of Days', () => {
      expect(getDayEachDay({ start: new Date(2019, 10, 25), end: new Date(2019, 11, 1) })).toEqual([
        new Date(2019, 10, 25),
        new Date(2019, 10, 26),
        new Date(2019, 10, 27),
        new Date(2019, 10, 28),
        new Date(2019, 10, 29),
        new Date(2019, 10, 30),
        new Date(2019, 11, 1),
      ])
    })
  })

  describe('#getMonthWeeks', () => {
    describe('when no `minNumberOfWeeks` has been passed', () => {
      test('returns a weeks array', () => {
        expect(getMonthWeeks({ activeMonth: new Date(2019, 11, 14), weekStartsOn: 1 })).toEqual([
          { weekStart: new Date(2019, 10, 25), weekNumber: 48, weekEnd: new Date(2019, 11, 1) },
          { weekStart: new Date(2019, 11, 2), weekNumber: 49, weekEnd: new Date(2019, 11, 8) },
          { weekStart: new Date(2019, 11, 9), weekNumber: 50, weekEnd: new Date(2019, 11, 15) },
          { weekStart: new Date(2019, 11, 16), weekNumber: 51, weekEnd: new Date(2019, 11, 22) },
          { weekStart: new Date(2019, 11, 23), weekNumber: 52, weekEnd: new Date(2019, 11, 29) },
          { weekStart: new Date(2019, 11, 30), weekNumber: 1, weekEnd: new Date(2020, 0, 5) },
        ])
      })
    })

    // describe('when `minNumberOfWeeks` has been passed', () => {
    //   test('returns a weeks array', () => {
    //     expect(getMonthWeeks({activeMonth: new Date(2018, 0, 1), minNumberOfWeeks: 8, weekStartsOn: 1})).toEqual([
    //       {weekStart: '2019-10-28', weekNumber: 44},
    //       {weekStart: '2019-11-04', weekNumber: 45},
    //       {weekStart: '2019-11-11', weekNumber: 46},
    //       {weekStart: '2019-11-18', weekNumber: 47},
    //       {weekStart: '2019-11-25', weekNumber: 48},
    //       {weekStart: '2019-12-02', weekNumber: 49},
    //       {weekStart: '2019-12-09', weekNumber: 50},
    //       {weekStart: '2019-12-16', weekNumber: 51},
    //     ])
    //   })
    // })
  })
})
