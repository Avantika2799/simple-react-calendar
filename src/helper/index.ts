import formatDate from 'date-fns/format'

export const getISODate = (date: Date) => formatDate(date, 'YYYY-MM-DD')
export const getDayFormatted = (date: Date) => formatDate(date, 'D')
export const getMonthNameFormatted = (date: Date) => formatDate(date, 'MMMM YYYY')
