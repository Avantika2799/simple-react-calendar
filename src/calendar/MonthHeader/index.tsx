import React, { FC, ReactElement } from 'react'

import {
  HandleOnMonthChange,
  RenderPropsMonthHeader,
  RenderPropsMonthHeaderButton
} from '../../@types'
import { getMonthNameFormatted } from '../../helper'

type Props = {
  activeMonth: Date
  blockClassName: string
  customRender: RenderPropsMonthHeader
  getMonthNameFormatted: typeof getMonthNameFormatted
  renderMonthHeaderButton: RenderPropsMonthHeaderButton
  onMonthChange: HandleOnMonthChange
}

const MonthHeader: FC<Props> = ({
  activeMonth,
  blockClassName,
  customRender,
  onMonthChange,
  getMonthNameFormatted,
  renderMonthHeaderButton
}) => {
  return customRender({
    monthTitle: getMonthNameFormatted(activeMonth),
    ComponentButtonPrev: () => < />,
    ComponentButtonNext: ,
  })
