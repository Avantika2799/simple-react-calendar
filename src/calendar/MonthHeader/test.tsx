import { cleanup, fireEvent, render, RenderResult } from '@testing-library/react'
import React from 'react'

import { getDayIndexIsWeekend, getMonthFormattedName, getWeekDays } from '../helper'
import MonthHeader, { Props } from './MonthHeader'

afterEach(cleanup)

describe('MonthHeader', () => {
  let api: RenderResult,
    activeMonth: Date,
    mockedCustomRender: any,
    props: Props,
    weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6

  beforeEach(() => {
    activeMonth = new Date(2019, 10, 1)
    weekStartsOn = 1

    props = {
      blockClassName: 'test-class-name',
      activeMonth,
      getMonthFormattedName,
      getWeekDays: () => getWeekDays({ daysOfWeek: ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'], dayIndex: 1 }),
      getDayIndexIsWeekend: (index) =>
        getDayIndexIsWeekend({ dayIndex: index as 0 | 1 | 2 | 3 | 4 | 5 | 6, weekStartsOn }),
      handleOnSwitchMonth: jest.fn(),
      isMonthNextEnabled: true,
      isMonthPrevEnabled: false,
      showDaysOfWeek: true,
      showMonthName: true,
      showMonthNext: true,
      showMonthPrev: true,
    }
  })

  describe('action handlers', () => {
    beforeEach(() => {
      api = render(<MonthHeader {...props} />)
    })

    test('invoke `#isMonthNextEnabled`', () => {
      const { getByTitle } = api

      fireEvent.click(getByTitle(/Next Month/i))

      expect(props.handleOnSwitchMonth).toHaveBeenCalledTimes(1)
    })
  })

  describe('when render prop has not been passed', () => {
    beforeEach(() => {
      api = render(<MonthHeader {...props} />)
    })

    test('default render', () => {
      const { container } = api

      expect(container).toMatchSnapshot()
    })

    describe('when `showMonthPrev`, `showMonthName`, `showMonthNext` `showDaysOfWeek` props are `falsy`', () => {
      beforeEach(() => {
        props = {
          ...props,
          showMonthPrev: false,
          showMonthName: false,
          showMonthNext: false,
          showDaysOfWeek: false,
        }
        api = render(<MonthHeader {...props} />)
      })

      test('default render', () => {
        const { container } = api

        expect(container).toMatchSnapshot()
      })
    })
  })

  describe('when render prop has been passed', () => {
    beforeEach(() => {
      mockedCustomRender = jest.fn(() => null)

      props = {
        ...props,
        customRender: mockedCustomRender,
      }
      api = render(<MonthHeader {...props} />)
    })

    test('render prop has been called properly', () => {
      expect(props.customRender).toHaveBeenCalledTimes(1)
      expect(props.customRender).toHaveBeenCalledWith(props)
    })
  })
})
