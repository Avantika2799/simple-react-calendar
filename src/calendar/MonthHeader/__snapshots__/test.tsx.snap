// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthHeader when render prop has not been passed default render 1`] = `
<div>
  <header
    class="test-class-name-month_header"
  >
    <div
      class="test-class-name-month_header_wrapper"
    >
      <button
        class="test-class-name-header_button is-prev is-disabled"
        data-calendar-value="-1"
        disabled=""
        title="Previous Month"
        type="button"
      >
        &gt;
      </button>
      <h3
        class="test-class-name-month_header_title"
      >
        November 2019
      </h3>
      <button
        class="test-class-name-header_button is-next"
        data-calendar-value="1"
        title="Next Month"
        type="button"
      >
        &lt;
      </button>
    </div>
    <div
      class="test-class-name-month_header_days"
    >
      <div
        class="test-class-name-days_of_week_day is-weekend"
      >
        Mon
      </div>
      <div
        class="test-class-name-days_of_week_day"
      >
        Tue
      </div>
      <div
        class="test-class-name-days_of_week_day"
      >
        Wed
      </div>
      <div
        class="test-class-name-days_of_week_day"
      >
        Thu
      </div>
      <div
        class="test-class-name-days_of_week_day"
      >
        Fri
      </div>
      <div
        class="test-class-name-days_of_week_day"
      >
        Sat
      </div>
      <div
        class="test-class-name-days_of_week_day is-weekend"
      >
        Sun
      </div>
    </div>
  </header>
</div>
`;

exports[`MonthHeader when render prop has not been passed when \`showMonthPrev\`, \`showMonthName\`, \`showMonthNext\` \`showDaysOfWeek\` props are \`falsy\` default render 1`] = `<div />`;
